---
title: "RNA-seq exploratory data analysis for INCLIVA cohort and i3s samples"
date: "`r Sys.Date()`"
author: "Arturo González-Vilanova"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: true
    highlight: tango
    embed_fonts: true
    number_sections: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
	echo = FALSE,
	fig.path = "../analysis/i3s_stad_immune_state_files/",
	message = FALSE,
	warning = FALSE,
	cache = TRUE,
	cache.path = "../analysis/i3s_stad_immune_state_cache/",
	collapse = TRUE,
	comment = "#>"
)
library(here)
library(data.table)
library(tximeta)
library(DESeq2)
library(ggplot2)
library(rtracklayer)
library(pheatmap)
library(RColorBrewer)
library(glmpca)
library(PoiClaClu)
library(sva)
```

```{r loading_coldata}
gtf_path <- here(file.path("analysis", "genome", "GCA_000001405.15_GRCh38_no_alt_analysis_set_genes.gtf"))
sample_features <- fread(here(file.path("scripts","sample_features.tsv")))[(rna == "Y" & (QC == "Passed" | is.na(publicadas)))]
files <- here(file.path("analysis","salmon",sample_features$names,"quant.sf"))
sample_features$files <- files
sample_features$batch <- as.factor(sample_features$batch)
ntpm <- 1 ## Minimum TMP for a gene to be considered expressed
```

```{r loading_expression_data}
suppressPackageStartupMessages(library(tximeta))
makeLinkedTxome(indexDir=here(file.path("analysis", "genome", "index", "salmon")),
                source="NCBI RefSeq",
                release="GCF_000001405.26_GRCh38/GRCh38_major_release_seqs_for_alignment_pipelines",
                organism="Homo sapiens",
                genome="GRCh38",
                fasta=here(file.path("analysis", "genome", "GCA_000001405.15_GRCh38_no_alt_analysis_set.fa")),
                gtf=gtf_path)
se <- summarizeToGene(tximeta(sample_features))

gtf <- as.data.table(import(gtf_path))[type == "gene", c("gene_id", "gene_biotype")]
rowData(se)$gene_biotype <- gtf$gene_biotype[match(rowData(se)$gene_id, gtf$gene_id)]
```

The total number of samples included in this analysis is `r nrow(sample_features)`. The INCLIVA cohort samples (identifier prefix "GTT") were included in the analysis given their RNA-seq data availability as well as appearance in the publication [Integrative immune transcriptomic classification improves patient selection for precision immunotherapy in advanced gastro-oesophageal adenocarcinoma](https://doi.org/10.1038/s41416-022-02005-z). Additionally, samples <mark>GTT33, GTT48 and GTT63</mark> where included for the specific purpose of performing batch correction with an adequate number of samples per batch.

# Number of expressed genes (TPM > 1)

```{r expressed_genes}
ngenes <- apply(assay(se, "abundance"), 2, function(x) sum(x  > ntpm))
sample_features$ngenes <- ngenes

ggplot(sample_features, aes(x = reorder(names, -ngenes), y = ngenes, fill = batch)) +
  geom_bar(position=position_dodge(), stat="identity") +
  labs(x = "Sample", y = "Number of genes > 1 TPM") +
  scale_fill_discrete(name = "Batch") +
  coord_flip()
```

# Expression by gene biotype

```{r expr_by_biotype}
min_tpm <- 10000

quant_abundance <- data.table(assay(se, "abundance"), keep.rownames = "gene_id")
quant_abundance$Biotype <- rowData(se)$gene_biotype
quant_abundance <- melt.data.table(quant_abundance, id.vars = c("gene_id","Biotype"), value.name = "TPM", variable.name = "Sample")
low_expr_biotypes <- quant_abundance[, min_tpm > sum(TPM), by = Biotype][V1 == T, Biotype]
quant_abundance[Biotype %in% low_expr_biotypes, Biotype := "Other"]
quant_abundance <- quant_abundance[,list(TPM = sum(TPM)), by = list(Sample,Biotype)]

my_colors <- scales::hue_pal()(length(levels(se$batch)))
names(my_colors) <- levels(se$batch)

ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = Biotype)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[se$batch]))
```

# Expression by chromosome

```{r expr_by_chr}
quant_abundance <- data.table(assay(se, "abundance"), keep.rownames = "gene_id")
quant_abundance$Chr <- as.data.frame(rowRanges(se))$seqnames
quant_abundance <- quant_abundance[Chr %in% paste0("chr",c(1:22,c("X","Y","M")))]
quant_abundance <- melt(quant_abundance, id.vars = c("gene_id","Chr"), value.name = "TPM", variable.name = "Sample")
quant_abundance <- quant_abundance[, list(TPM = sum(TPM)), by = list(Sample,Chr)]
quant_abundance[, ChrM := ifelse(Chr == "chrM","chrM","Other")]

ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = Chr)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[se$batch])) +
  labs(fill=NULL)
```

```{r expr_by_chrM}
quant_abundance <- quant_abundance[,.(TPM=sum(TPM)), by = .(Sample,ChrM)]
ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = ChrM)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[se$batch])) +
  labs(fill=NULL)
```

```{r filter}
#keep <- rowSums(assay(se, "abundance") >= ntpm) >= min(table(se$batch))
keep <- rowSums(assay(se, "counts")) >= 1
se <- se[keep,]
```

# Generalized PCA

```{r generalized_pca}
gpca <- glmpca(assay(se, "counts"), L=2, fam = "nb")
gpca.dat <- gpca$factors
gpca.dat$batch <- se$batch
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = batch)) +
  geom_point(size = 3) +
  labs(x = "PC1", y = "PC2", color = "Batch") +
  coord_fixed() +
  ggtitle("Generalized PCA")
```

# Distance matrix

```{r heatmap}
poisd <- PoissonDistance(t(assay(se, "counts")))
attr(poisd$dd, "Labels") <- se$names
samplePoisDistMatrix <- as.matrix(poisd$dd)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotation <- as.data.frame(sample_features[,"batch",drop=F])
rownames(annotation) <- sample_features$names
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         annotation_col=annotation,
         annotation_colors = list(batch = my_colors),
         col = colors,
         annotation_names_col = F,
         angle = 90,
         treeheight_row = 0)
```

# Batch adjustment by ComBat-seq

```{r batch_adjustment, echo = T}
se_adjusted_counts <- ComBat_seq(assay(se,"counts"), batch=se$batch, group=NULL)
vsd_adjusted <- vst(round(se_adjusted_counts))
tpm_adjusted <- t(t(se_adjusted_counts*10**6/assay(se,"length"))/colSums(se_adjusted_counts/assay(se,"length")))
```
## Generalized PCA

```{r generalized_pca_adjusted}
gpca <- glmpca(se_adjusted_counts, L=2, fam = "nb")
gpca.dat <- gpca$factors
gpca.dat$batch <- se$batch
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = batch)) +
  geom_point(size = 3) +
  labs(x = "PC1", y = "PC2", color = "Batch") +
  coord_fixed() +
  ggtitle("Generalized PCA")
```

## Distance matrix

```{r heatmap_adjusted}
poisd <- PoissonDistance(t(se_adjusted_counts))
attr(poisd$dd, "Labels") <- se$names
samplePoisDistMatrix <- as.matrix(poisd$dd)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotation <- as.data.frame(sample_features[,"batch",drop=F])
rownames(annotation) <- sample_features$names
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         annotation_col=annotation,
         annotation_colors = list(batch = my_colors),
         annotation_names_col = F,
         col = colors,
         angle = 90,
         treeheight_row = 0)
```
```{r write_xlsx}
library(openxlsx)
wb <- createWorkbook(creator="Arturo González-Vilanova",
                     title="RNA-seq quantification data (18/04/2023)")
addWorksheet(wb, "VST batch adjusted")
writeData(wb, sheet = "VST batch adjusted", rowNames = T,
          x = vsd_adjusted[,-match(c("GTT33","GTT48","GTT63"),colnames(vsd_adjusted))])
addWorksheet(wb, "TPM batch adjusted")
writeData(wb, sheet = "TPM batch adjusted", rowNames = T,
          x = tpm_adjusted[,-match(c("GTT33","GTT48","GTT63"),colnames(tpm_adjusted))])
addWorksheet(wb, "Counts batch adjusted")
writeData(wb, sheet = "Counts batch adjusted", rowNames = T,
          x = se_adjusted_counts[,-match(c("GTT33","GTT48","GTT63"),colnames(se_adjusted_counts))])
saveWorkbook(wb, here(file.path("analysis","rnaseq_quant_incliva_i3s.xlsx")), overwrite = TRUE)
```

# Immune signatures

## Heatmaps at the immune signature level

```{r load_signatures}
signatures <- fread(here(file.path("scripts","immune_functional_signatures.tsv")), na.strings="")
immune_genes <- unique(unlist(lapply(signatures, na.omit)))
vsd_immune <- vsd_adjusted[immune_genes,]
vsd_immune <- vsd_immune[,-match(c("GTT33","GTT48","GTT63"),colnames(vsd_immune))]
```

```{r is_heatmaps, fig.height = 9}
hmap_list <- lapply(names(signatures), function(signature){
  genes <- na.omit(signatures[[signature]])
  quant <- vsd_immune[genes,]
  hmap <- pheatmap(quant,
                   fontsize_row= 10 - nrow(quant) / 15,
                   angle = 45,
                   main = signature,
                   cluter_rows = T,
                   cluster_cols = T,
                   clustering_method = "ward.D",
                   annotation_names_col = F,
                   cutree_cols = 2)
  return(hmap)
})
names(hmap_list) <- names(signatures)
```

## Integrative heatmap

```{r heatmap_classification}
immune_classes <- data.frame(lapply(hmap_list, function(hmap){ cutree(hmap$tree_col, k = 2) }), check.names = F)
annotation$Cohort <- ifelse(grepl("^S",rownames(annotation)),"i3S","INCLIVA") 
pheatmap(t(immune_classes),
         annotation_col = annotation[colnames(vsd_immune),"Cohort",drop=F],
         angle = 45,
         clustering_method = "ward.D",
         legend = F,
         annotation_legend = F,
         treeheight_row = 0,
         cutree_cols = 2)
```





