---
title: "RNA-seq exploratory data analysis for INCLIVA cohort and i3s samples"
date: "`r Sys.Date()`"
author: "Arturo GonzÃ¡lez-Vilanova"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: true
    highlight: tango
    embed_fonts: true
    number_sections: true
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = FALSE,
  message = FALSE,
  warning = FALSE,
  cache = TRUE
)
library(here)
library(data.table)
library(tximeta)
library(DESeq2)
library(ggplot2)
library(rtracklayer)
library(pheatmap)
library(RColorBrewer)
library(glmpca)
library(PoiClaClu)
library(sva)
```

```{r loading_coldata}
gtf_path <- here(file.path("analysis", "genome", "GCA_000001405.15_GRCh38_no_alt_analysis_set_genes.gtf"))
sample_features <- fread(here(file.path("scripts","sample_features.tsv")))[(rna == "Y" & (QC == "Passed" | is.na(publicadas)))]
files <- here(file.path("analysis","salmon",sample_features$names,"quant.sf"))
sample_features$files <- files
sample_features$batch <- as.factor(sample_features$batch)
```

```{r loading_expression_data}
suppressPackageStartupMessages(library(tximeta))
makeLinkedTxome(indexDir=here(file.path("analysis", "genome", "index", "salmon")),
                source="NCBI RefSeq",
                release="GCF_000001405.26_GRCh38/GRCh38_major_release_seqs_for_alignment_pipelines",
                organism="Homo sapiens",
                genome="GRCh38",
                fasta=here(file.path("analysis", "genome", "GCA_000001405.15_GRCh38_no_alt_analysis_set.fa")),
                gtf=gtf_path)
se <- summarizeToGene(tximeta(sample_features))
dds <- DESeqDataSet(se, design = ~ batch)

gtf <- as.data.table(import(gtf_path))[type == "gene", c("gene_id", "gene_biotype")]
rowData(dds)$gene_biotype <- gtf$gene_biotype[match(rowData(dds)$gene_id, gtf$gene_id)]
```

The total number of samples included in this analysis is `r nrow(sample_features)`. The INCLIVA cohort samples (identifier prefix "GTT") were included in the analysis given their RNA-seq data availability as well as appearance in the publication [Integrative immune transcriptomic classification improves patient selection for precision immunotherapy in advanced gastro-oesophageal adenocarcinoma](https://doi.org/10.1038/s41416-022-02005-z). Additionally, samples <mark>GTT33, GTT48 and GTT63</mark> where included for the specific purpose of performing batch correction with an adequate number of samples per batch.

# Number of expressed genes (TPM > 1)

```{r expressed_genes}
ntpm <- 1
ngenes <- apply(assay(dds, "abundance"), 2, function(x) sum(x  > ntpm))
sample_features$ngenes <- ngenes

ggplot(sample_features, aes(x = reorder(names, -ngenes), y = ngenes, fill = batch)) +
  geom_bar(position=position_dodge(), stat="identity") +
  labs(x = "Sample", y = "Number of genes > 1 TPM") +
  scale_fill_discrete(name = "Batch") +
  coord_flip()
```

# Expression by gene biotype

```{r expr_by_biotype}
min_count <- 10000

quant_abundance <- data.table(assay(dds, "abundance"), keep.rownames = "gene_id")
quant_abundance$Biotype <- rowData(dds)$gene_biotype
quant_abundance <- melt.data.table(quant_abundance, id.vars = c("gene_id","Biotype"), value.name = "TPM", variable.name = "Sample")
low_expr_biotypes <- quant_abundance[, min_count > sum(TPM), by = Biotype][V1 == T, Biotype]
quant_abundance[Biotype %in% low_expr_biotypes, Biotype := "Other"]
quant_abundance <- quant_abundance[,list(TPM = sum(TPM)), by = list(Sample,Biotype)]

my_colors <- scales::hue_pal()(length(levels(dds$batch)))
names(my_colors) <- levels(dds$batch)

ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = Biotype)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[dds$batch]))
```

# Expression by chromosome

```{r expr_by_chr}
quant_abundance <- data.table(assay(dds, "abundance"), keep.rownames = "gene_id")
quant_abundance$Chr <- as.data.frame(rowRanges(dds))$seqnames
quant_abundance <- quant_abundance[Chr %in% paste0("chr",c(1:22,c("X","Y","M")))]
quant_abundance <- melt(quant_abundance, id.vars = c("gene_id","Chr"), value.name = "TPM", variable.name = "Sample")
quant_abundance <- quant_abundance[,list(TPM = sum(TPM)), by = list(Sample,Chr)]
quant_abundance <- quant_abundance[,ChrM := ifelse(Chr == "chrM","chrM","Other")]

ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = Chr)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[dds$batch])) +
  labs(fill=NULL)
```

```{r expr_by_chrM}
quant_abundance[,sum(TPM), by = .(Sample,ChrM)]
ggplot(quant_abundance, aes(x = Sample, y = TPM, fill = ChrM)) +
  geom_bar(stat="identity") +
  coord_flip() +
  theme(axis.text.y = element_text(colour = my_colors[dds$batch])) +
  labs(fill=NULL)
```

```{r filter}
keep <- rowSums(assay(dds, "abundance") >= 1) >= min(table(dds$batch))
dds <- dds[keep,]
```

# Generalized PCA

```{r generalized_pca}
gpca <- glmpca(counts(dds), L=2, fam = "nb")
gpca.dat <- gpca$factors
gpca.dat$batch <- dds$batch
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = batch)) +
  geom_point(size = 3) +
  labs(x = "PC1", y = "PC2", color = "Batch") +
  coord_fixed() +
  ggtitle("Generalized PCA")
```

# Distance matrix

```{r heatmap}
poisd <- PoissonDistance(t(counts(dds)))
attr(poisd$dd, "Labels") <- dds$names
samplePoisDistMatrix <- as.matrix(poisd$dd)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotation <- as.data.frame(sample_features[,"batch",drop=F])
rownames(annotation) <- sample_features$names
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         annotation_col=annotation,
         annotation_colors = list(batch = my_colors),
         col = colors,
         angle = 90)
```

# Batch adjustment by ComBat-seq

```{r batch_adjustment, echo = T}
dds_adjust <- dds
dds_adjusted_counts <- ComBat_seq(counts(dds), batch=dds$batch, group=NULL)
```
## Generalized PCA

```{r generalized_pca_adjusted}
gpca <- glmpca(dds_adjusted_counts, L=2, fam = "nb")
gpca.dat <- gpca$factors
gpca.dat$batch <- dds$batch
ggplot(gpca.dat, aes(x = dim1, y = dim2, color = batch)) +
  geom_point(size = 3) +
  labs(x = "PC1", y = "PC2", color = "Batch") +
  coord_fixed() +
  ggtitle("Generalized PCA")
```

## Distance matrix

```{r heatmap_adjusted}
poisd <- PoissonDistance(t(dds_adjusted_counts))
attr(poisd$dd, "Labels") <- dds$names
samplePoisDistMatrix <- as.matrix(poisd$dd)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
annotation <- as.data.frame(sample_features[,"batch",drop=F])
rownames(annotation) <- sample_features$names
pheatmap(samplePoisDistMatrix,
         clustering_distance_rows = poisd$dd,
         clustering_distance_cols = poisd$dd,
         annotation_col=annotation,
         col = colors,
         angle = 90)
```

# Immune signatures

```{r load_signatures}
signatures <- fread(here(file.path("scripts","immune_functional_signatures.tsv")), na.strings="")
signatures <- lapply(signatures, na.omit)
```




